File: core/OneAwayLcci/OneAwayLcci.java
package com.dbc;

public class OneAwayLcci {
    public boolean oneEditAway(String first, String second) {
        int[][] dp = new int[first.length() + 1][second.length() + 1];
        for (int i = 0; i < first.length() + 1; i++) dp[i][0] = i;
        for (int i = 0; i < second.length() + 1; i++) dp[0][i] = i;
        for (int i = 1; i < first.length() + 1; i++) {
            for (int j = 1; j < second.length() + 1; j++) {
                int left = dp[i][j - 1] + 1;
                int down = dp[i - 1][j] + 1;
                int leftDown = dp[i - 1][j - 1];
                if (first.charAt(i - 1) != second.charAt(j - 1)) leftDown++;
                dp[i][j] = Math.min(left, Math.min(down, leftDown));
            }
        }
        return dp[first.length()][second.length()] == 0 || dp[first.length()][second.length()] == 1;
    }

    public boolean oneEditAway1(String first, String second) {
        if (Math.abs(first.length() - second.length()) > 1) return false;
        if (first.length() < second.length()) return oneEditAway1(second, first);
        for (int i = 0; i < second.length(); i++) {
            if (first.charAt(i) != second.charAt(i)) {
                if (first.length() == second.length()) return first.substring(i + 1).equals(second.substring(i + 1));
                else return first.substring(i + 1).equals(second.substring(i));
            }
        }
        return true;
    }
}

File: core/OneAwayLcci/OneAwayLcci.py
def oneEditAway(first: str, second: str) -> bool:
    if first == "" or second == "":
        return (first == "" and len(second) < 2) or (second == "" and len(first) < 2)
    len1, len2 = len(first), len(second)
    dp = [[0] * (len2 + 1) for _ in range(len1 + 1)]
    for index in range(len2 + 1):
        dp[0][index] = index
    for index in range(len1 + 1):
        dp[index][0] = index

    for row in range(1, len1 + 1):
        for col in range(1, len2 + 1):
            left = dp[row][col - 1] + 1
            down = dp[row - 1][col] + 1
            left_down = dp[row - 1][col - 1]

            if first[row - 1] != second[col - 1]:
                left_down += 1

            dp[row][col] = min(left, min(down, left_down))

    return dp[-1][-1] == 0 or dp[-1][-1] == 1


if __name__ == '__main__':
    print(oneEditAway("pale", "ple"))
